Add Get-Uptime

iSazonov opened the pull request:
1. Add Get-Uptime into Microsoft.PowerShell.Utility.psm1
2. Add alias "uptime"
3. Correct typo
4. Add test

Close #2497

iSazonov committed:
Add Get-Uptime

1. Add Get-Uptime into Microsoft.PowerShell.Utility.psm1
2. Add alias "uptime"
3. Correct typo
4. Add test

msftclas commented:
Hi __@iSazonov__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

rkeithhill commented on src/Modules/Windows-Core/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1:
```diff
@@ -24,8 +24,8 @@ CmdletsToExport= "Format-List", "Format-Custom", "Format-Table", "Format-Wide",
-    "ConvertFrom-SddlString"
-AliasesToExport= "fhx"
+    "ConvertFrom-SddlString", "Get-Uptime"
+AliasesToExport= "fhx", "uptime"
```
Haven't we been through this before with ls and curl i.e. superseding a native utility with a PowerShell alias.  I wouldn't define this alias on *nix.  Windows perhaps but not *nix.

oising replied:
yep, and you can still invoke it as `uptime` on systems without a native version, due to the last-ditch command search dropping the verb and searching for noun for "get" commands.

iSazonov replied:
@oising Thanks for comment! I'll see how it works.

mirichmo commented:
@JamesWTruher Please review

JamesWTruher commented on src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1:
```diff
@@ -737,4 +737,18 @@ function ConvertFrom-SddlString
+    Process
+    {
+        # It is portable
+        [TimeSpan]::FromSeconds([system.diagnostics.stopwatch]::GetTimestamp()/[system.diagnostics.stopwatch]::Frequency)
```
>GetTimestamp [](start = 64, length = 12)

I believe this API is hardware dependent - if `IsHighResolution` false, you'll get back `[datetime]::now.ticks` which won't give the proper answers. You might want to be checking for that. something like:
```
if ( [system.diagnostics.stopwatch]::IsHighResolution) 
{
[TimeSpan]::FromSeconds([system.diagnostics.stopwatch]::GetTimestamp()/[system.diagnostics.stopwatch]::Frequency)
}
. . .
```

iSazonov replied:
@JamesWTruher    Is there universal (portable) way to calculate uptime?

I agree with "if IsHighResolution" but I do not like "throw". Really today are there systems where it is possible? Perhaps a better return $null?

daxian-dbw replied:
+1 for returning `null` in that case.

JamesWTruher commented:
:clock1:

daxian-dbw commented on src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1:
```diff
@@ -737,4 +737,18 @@ function ConvertFrom-SddlString
+function Get-Uptime
+{
+    [CmdletBinding()]
+    [Alias("uptime")]
```
The alias collides with the native utility `/usr/bin/uptime` in Linux. I don't think we want PS cmdlets to hide the native utility (e.g. `Invoke-WebRequest` had the alias `curl` and `wget`, and it caused problems in Linux PS), so we should avoid using this alias.

oising replied:
There's an RFC for "weak aliases" that should cover this. Currently, aliases are the top priority, ahead of native commands. The weak alias spec allows for aliases that have the lowest priority, so a native uptime would always win. Another way to do this is to not use an alias at all: Just leave it as `get-uptime`. Systems that don't have a native `uptime` will yield to Powershell's last-ditch command search, which strips the get verb and tries to look for the noun. Systems with a native uptime will be found before the last-ditch noun search kicks in.

iSazonov replied:
@oising I don't like "low priority for alias", because it may cause unpredictable behavior 😕

rkeithhill replied:
BTW the "weak aliases" RFC was rejected.

lzybkr commented on src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1:
```diff
@@ -737,4 +737,18 @@ function ConvertFrom-SddlString
\ No newline at end of file
+}
+
+function Get-Uptime
```
I'd prefer to see this as a cmdlet in C#.

For faster startup, we need to avoid PowerShell script in the core modules that are loaded in essentially every session.

iSazonov replied:
@rkeithhill @JamesWTruher  @daxian-dbw if you approve this I'll reopen PR with C#.

daxian-dbw commented on src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1:
```diff
@@ -432,9 +432,9 @@ function Format-Hex
-                    # 00000020   B2 00 00 00 B3 00 00 00 B4 00 00 00              �...�...�...
+                    # 00000000   AA 00 00 00 AB 00 00 00 AC 00 00 00 AD 00 00 00  ª...«...¬...­...
+                    # 00000010   AE 00 00 00 AF 00 00 00 B0 00 00 00 B1 00 00 00  ®...¯...°...±...
+                    # 00000020   B2 00 00 00 B3 00 00 00 B4 00 00 00              ²...³...´...
```
It looks like you are changing the format (indentation and etc.) in this files. The guideline in [coding conventions](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/coding-guidelines.md#coding-conventions) suggests to avoid reformatting any code when submitting a PR. Could you please revert those formatting changes?

iSazonov replied:
daxian-dbw Sorry. Here is conflict VS Code Editor with special chars. I tried to fix this in other editors but unsuccessfully.
Fortunately this is just a comment. Can remove them?

daxian-dbw commented on src/Modules/Windows-Full/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1:
```diff
@@ -2,7 +2,7 @@
 Author="Microsoft Corporation"
 CompanyName="Microsoft Corporation"
-Copyright="� Microsoft Corporation. All rights reserved."
+Copyright="© Microsoft Corporation. All rights reserved."
```
same here. Please avoid formatting changes.

iSazonov replied:
Again VS Code. See comment above.

daxian-dbw commented:
@iSazonov also, I found you didn't update the unix version of `utility.psd1` (`Unix\Microsoft.PowerShell.Utility\Microsoft.PowerShell.Utility.psd1`), so is it intentional to not expose this cmdlet in Linux?

iSazonov committed:
Fix after review

1. Remove alias
2. Add IsHighResolution check and throw
3. Modify .psd1 for unix

iSazonov commented:
daxian-dbw  utility.psd1  fixed

iSazonov commented:
😕 What's the problem with tests? Second PR breaks down...

vors commented:
@iSazonov it's unrelated to your changes. Same story for the build from `master` https://ci.appveyor.com/project/PowerShell/powershell/build/6.0.0-alpha.11-1017#L386

It's likely changes in `dotnet` cli. We will try to resolve it today.

msftclas commented:
Hi __@iSazonov__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

vors commented:
Reopen PR to build on the newer version of master (that includes a fix for new `dotnet`)

iSazonov commented:
I close the PR to continue with #2517
